streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -583,6 +627,172 @@ public void advanceWallClockTime(final long advanceMs) {
+                                                                final Deserializer<K> keyDeserializer,
+                                                                final Deserializer<V> valueDeserializer) {
+        return new TestOutputTopic<K, V>(this, topicName, keyDeserializer, valueDeserializer);
+    }
+
+    /**
+     * Read the next record from the given topic.
+     * These records were output by the topology during the previous calls to {@link #pipeInput(ConsumerRecord)}.
+     *
+     * @param topic the name of the topic
+     * @return the next record on that topic, or {@code null} if there is no record available
+     */
+    @SuppressWarnings("WeakerAccess")
+    ProducerRecord<byte[], byte[]> readRecord(final String topic) {
+        final Queue<? extends ProducerRecord<byte[], byte[]>> outputRecords = getRecordsQueue(topic);
+        if (outputRecords == null) {
+            return null;
+        }
+        return outputRecords.poll();
+    }
+
+    /**
+     * Read the next record from the given topic.
+     * These records were output by the topology during the previous calls to {@link #pipeInput(ConsumerRecord)}.
+     *
+     * @param topic             the name of the topic
+     * @param keyDeserializer   the deserializer for the key type
+     * @param valueDeserializer the deserializer for the value type
+     * @param <K> the key type
+     * @param <V> the value type
+     * @return the next record on that topic, or {@code null} if there is no record available
+     */
+    @SuppressWarnings("WeakerAccess")
+    <K, V> TestRecord<K, V> readRecord(final String topic,
+                                         final Deserializer<K> keyDeserializer,
+                                         final Deserializer<V> valueDeserializer) {
+        final Queue<? extends ProducerRecord<byte[], byte[]>> outputRecords = getRecordsQueue(topic);
+        if (outputRecords == null) {
+            throw new NoSuchElementException("Uninitialized topic: " + topic);
+        }
+        final ProducerRecord<byte[], byte[]> record = outputRecords.poll();
+        if (record == null) {
+            throw new NoSuchElementException("Empty topic: " + topic);
+        }
+        final K key = keyDeserializer.deserialize(record.topic(), record.key());
+        final V value = valueDeserializer.deserialize(record.topic(), record.value());
+        return new TestRecord<>(key, value, record.headers(), record.timestamp());
+    }
+
+    /**
+     * Serialize an input recond and send on the specified topic to the topology and then
+     * commit the messages.
+     *
+     * @param topic           the name of the topic
+     * @param record          TestRecord to be send
+     * @param keySerializer   the key serializer
+     * @param valueSerializer the value serializer
+     * @param <K> the key type
+     * @param <V> the value type*
+     * @param time            timestamp to override the record timestamp
+     */
```

Comment:
we can omit javadoc for non-public methods

Reply:
removed